x-pack/legacy/plugins/infra/public/containers/source_id/source_id.ts

```diff
@@ -25,4 +28,7 @@ export const replaceSourceIdInQueryString = (sourceId: string) =>
 const sourceIdRuntimeType = runtimeTypes.union([runtimeTypes.string, runtimeTypes.undefined]);
 const encodeSourceIdUrlState = sourceIdRuntimeType.encode;
 const decodeSourceIdUrlState = (value: unknown) =>
-  sourceIdRuntimeType.decode(value).getOrElse(undefined);
+  pipe(
+    sourceIdRuntimeType.decode(value),
+    fold(constant(undefined), identity)
```

Comment:
Unfortunately the type definition of `getOrElse` requires the return value to be of the same type as the right value of the data type. E.g., if we return undefined, TS will complain about the right value of `sourceIdRunTimeType.decode(value)` not being of type `undefined`:

![image](https://user-images.githubusercontent.com/352732/64429216-ca1bfa80-d0b5-11e9-8988-093857f14008.png)

It's unfortunate because IMHO it severely limits the usefulness of `getOrElse`. It seems strange, I might be doing something wrong, so if you have any suggestions let me know.

A workaround could be a type assertion, but that's not particularly safe.

![image](https://user-images.githubusercontent.com/352732/64429359-18c99480-d0b6-11e9-9053-84023b7f0fc5.png)

I think the best way to solve this right now is to create a utility (maybe `getOrDefault`) that uses `fold` under the hood, but that's probably better left to the teams themselves to figure out.

Reply:
thanks for the explanation, let's keep `fold` for now